### DISCUSSION TITLE
New unified expression syntax

### DISCUSSION THREAD

**Discussion by tmedwards** [tmedwards/sugarcube-3#32] [Ideas]
I'm thinking about a new unified syntax to replace the various existing "expression" syntaxes: the naked variable markup, the `<<print>>` macro, macro argument backquote expressions, and the HTML evaluation attribute directive (`@attr`).

Having a single unified syntax for evaluating an expression and yielding its value would be good.  There would probably still be a least a couple of actual mechanisms under the hood, but the user-facing syntax itself would be the same.

The syntax I'm currently mulling over, though that's not set in stone, is the double curly braces (`{{…}}`).

Aside from unifying an expression syntax, this would also free up backquotes so you may use template literals in macro argument expressions.  This wasn't an issue for v2 considering its browser support pledge, but v3's support target is currently ES2020 and template literals are too nice to continue to co-opt.

### Naked variable markup &  `<<print>>`

```
$foo
<<= $foo>>
    vs.
{{$foo}}

$foo.method()  /* ← INVALID */
<<= $foo.method()>>
    vs.
{{$foo.method()}}
```

### Macro argument backquote expressions

```
<<link "Do the thing" `some expression`>>…
    vs.
<<link "Do the thing" {{some expression}}>>…
```

### HTML evaluation attribute directive

```
<span @someattrib="some expression">…

<span someattrib="{{some expression}}">…
```

**Comment by ChapelR** [comment]
I like this idea. 

Are you planning to remove naked variables though? I feel like you could keep it and still transition everything else to the unified markup here.

**Reply by tmedwards** [reply]
Hadn't gotten that far in my thinking.  One the one hand needing mustaches, or whatever it ends up being, (e.g., `{{$foo}}`) isn't the end of the world and keeps things neat by having a single "print" syntax in this context.  On the other, being able to just write `$foo` is nice.

Even if I did decide to nix the naked variable markup, and possibly `<<print>>`, I'd just deprecate it/them, so they'd still work in v3.

I don't feel too strongly either way at the moment, but the thought of replacing NVM and `<<print>>` with a single syntax is kind of nice.

**Reply by greyelf** [reply]
One issue that keeps coming up with naked variables is authors trying to output an expression more complex than what they are designed to handle. eg. $list[_index].doit($var)

While I do like the easy of being able to write `$variable` to output the value I think that if naked variables are still a thing in SC3 then you will need to make it very clear on exactly just how limited their syntax is. Personally I would consider making naked variables deprecated.

**Reply by tmedwards** [reply]
Having thought about this some more, I think what I want to do is keep the NVM but restrict it to only simple variable access—i.e., `$foo` only, no `$foo[…]` or `$foo.…`.

The `{{…}}` syntax would handle everything else.

**Comment by HiEv** [comment]
While I generally like it, I feel like eventually someone is going to run into a case where they're mixing objects `{ ... }`, inline or block code markup `{{{ ... }}}`, or text strings with curly brackets in them with the `{{ ... }}` markup in a way that breaks things.

But, then again, that's probably the case regardless of what standard keyboard character you use.  Anyways, just something to keep in mind.
